<a href="https://colab.research.google.com/github/DipeshChothwe/NLP/blob/master/Sentiment_analysis_for_world_leaders.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [2]:
!pip install GetOldTweets3
import string
from collections import Counter

import matplotlib.pyplot as plt

In [3]:
#function for scrapping tweets 

def get_tweets(date1, date2, tweetname):
    import GetOldTweets3 as got   # library used to scrape data from twitter without any other tools
    tweetCriteria = got.manager.TweetCriteria().setQuerySearch(tweetname) \
        .setSince(date1) \
        .setUntil(date2) \
        .setLang('en') \
        .setMaxTweets(3000)

    # Creation of list that contains all tweets
    tweets = got.manager.TweetManager.getTweets(tweetCriteria)

    # Creating list of chosen tweet data
    text_tweets = [[tweet.text] for tweet in tweets]
    return text_tweets

In [6]:
# Fetch tweets:

tweets_trump = get_tweets("2020-07-20", "2020-08-20", "@realDonaldTrump")
tweets_modi = get_tweets("2020-07-20", "2020-08-19", "@narendramodi")
tweets_joe = get_tweets("2020-07-20", "2020-08-20", "@JoeBiden")

In [4]:
#Defining stop words:

stop_words = {"i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself",
              "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself",
              "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these",
              "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do",
              "does", "did", "doing", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while",
              "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before",
              "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again",
              "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each",
              "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than",
              "too", "very", "s", "t", "can", "will", "just", "don", "should", "now"}

In [59]:
#preprocessing text for Bag of Words:


def preprocessing_BOW(list_msgs):
  msgs = ""
  for i in range(0, len(list_msgs)):
    msgs = list_msgs[i][0] + " " + msgs
  
  # 1)Change letters to lowercase; 2)Remove punctuactions; 3)Split sentence in words:
  a = msgs.lower()
  cleaned_words = a.translate(str.maketrans('', '', string.punctuation))
  tokanized_words = cleaned_words.split()

  #Tokanization:
  tokanized_words_without_stopwords = set(tokanized_words)
  tokanized_words_without_stopwords.difference_update(stop_words)

  return (tokanized_words, cleaned_words)

In [60]:

tokanized_words_without_stopwords_trump, cleaned_words_trump = preprocessing_BOW(tweets_trump)
tokanized_words_without_stopwords_modi, cleaned_words_modi = preprocessing_BOW(tweets_modi)
tokanized_words_without_stopwords_joe, cleaned_words_joe = preprocessing_BOW(tweets_joe)

In [48]:
# Get emotions text
emotion_list_trump = []
emotion_list_modi = []
emotion_list_joe = []
Word_not_emmotions = []
word_to_emotions = {}


with open('/content/drive/My Drive/NLP/emotions.txt', 'r') as file:
    for line in file:
        clear_line = line.replace('\n', '').replace(',', '').replace("'", '').strip()
        word, emotion = clear_line.split(':')

        if word in tokanized_words_without_stopwords_trump:
            emotion_list_trump.append(emotion)
            word_to_emotions[word] = emotion
        if word in tokanized_words_without_stopwords_modi:
            emotion_list_modi.append(emotion)
            word_to_emotions[word] = emotion
        if word in tokanized_words_without_stopwords_joe:
            emotion_list_joe.append(emotion)
            word_to_emotions[word] = emotion

trump = Counter(emotion_list_trump)
modi = Counter(emotion_list_modi)
joe = Counter(emotion_list_joe)

print("Emotions for Donald Trump:")
print(trump)
print("Emotions for Naredra Modi:")
print(modi)
print("Emotions for Joe Biden:")
print(joe)

Emotions for Donald Trump:
Counter({' happy': 14, ' sad': 13, ' fearful': 12, ' powerless': 9, ' angry': 8, ' attracted': 7, ' attached': 7, ' adequate': 6, ' cheated': 5, ' singled out': 5, ' fearless': 4, ' esteemed': 4, ' surprise': 3, ' free': 3, ' hated': 3, ' entitled': 3, ' safe': 2, ' average': 2, ' anxious': 2, ' embarrassed': 2, ' loved': 2, ' belittled': 1, ' ecstatic': 1, ' codependent': 1, ' apathetic': 1, ' lost': 1, ' obsessed': 1, ' independent': 1, ' derailed': 1, ' alone': 1, ' focused': 1})
Emotions for Naredra Modi:
Counter({' sad': 15, ' happy': 14, ' attracted': 10, ' attached': 9, ' fearful': 9, ' powerless': 7, ' cheated': 5, ' esteemed': 5, ' fearless': 5, ' adequate': 5, ' angry': 4, ' surprise': 3, ' free': 3, ' alone': 3, ' hated': 3, ' anxious': 3, ' loved': 3, ' average': 2, ' focused': 2, ' singled out': 2, ' bored': 1, ' safe': 1, ' demoralized': 1, ' ecstatic': 1, ' lost': 1, ' codependent': 1, ' independent': 1, ' embarrassed': 1, ' entitled': 1, ' apa

In [70]:
import nltk
nltk.download("vader_lexicon")

from nltk.sentiment.vader import SentimentIntensityAnalyzer as SIA

def sentiment_analyse(sentiment_text):
    sid = SIA()
    score = sid.polarity_scores(sentiment_text)
    
    if score['neg'] > score['pos']:
        print("Negative Sentiment with positive to negative ratio: ", score['pos']/score['neg'])
        
    elif score['neg'] < score['pos']:
        print("Positive Sentiment with positive to negative ratio: ", score['pos']/score['neg'])
    else:
        print("Neutral Sentiment")
    return (score['pos']/score['neg'])

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [71]:

print("Sentiment for Donald Trump:")
trump_score = sentiment_analyse(cleaned_words_trump)
print("Sentiment for Narendra Modi:")
modi_score = sentiment_analyse(cleaned_words_modi)
print("Sentiment for Joe Biden:",)
joe_score = sentiment_analyse(cleaned_words_joe)


Sentiment for Donald Trump:
Negative Sentiment with positive to negative ratio:  0.9068322981366459
Sentiment for Narendra Modi:
Positive Sentiment with positive to negative ratio:  1.4954128440366974
Sentiment for Joe Biden:
Positive Sentiment with positive to negative ratio:  1.0921985815602837


In [76]:
print("Positive to negative sentiment ratio between Joe Biden and Donald Trump",)
print(joe_score/trump_score)

print("Percentage difference between positive to negative sentiment ratio: Joe Biden vs Donald Trump",)
print(((joe_score/trump_score) - 1)*100)

Positive to negative sentiment ratio between Joe Biden and Donald Trump
1.2044107645972992
Percentage difference between positive to negative sentiment ratio: Joe Biden vs Donald Trump
20.441076459729924


---
# **Conclusion:**

**From the tweets in the last months:**
*   Joe Biden has slightly positive sentiment towards him from the community.
*   Donald Trump has slightly negative sentiment towards him from the community.
*   Joe Biden has considerable 20.44% more positive sentiment with him from the communituty compared to Donald Trump.
*   Naredra Modi has been able to accumulate high positive sentiment towards him even in these difficult and continuosly worsening pandemic times.


---




# **Part 2**


*   Documentation of words and associated emotion:




In [85]:
print( "Word : assigned emotio") 
print("")
word_to_emotions

Word : assigned emotio



{'abandoned': ' hated',
 'abused': ' powerless',
 'accepted': ' loved',
 'accused': ' cheated',
 'adequate': ' adequate',
 'admired': ' esteemed',
 'affected': ' attracted',
 'afraid': ' fearful',
 'alone': ' alone',
 'angry': ' angry',
 'animated': ' happy',
 'anxious': ' attracted',
 'appalled': ' angry',
 'appreciated': ' esteemed',
 'arrogant': ' independent',
 'ashamed': ' embarrassed',
 'attached': ' attached',
 'attacked': ' fearful',
 'average': ' average',
 'awkward': ' embarrassed',
 'baffled': ' lost',
 'belonging': ' attached',
 'betrayed': ' cheated',
 'bitter': ' angry',
 'blocked': ' powerless',
 'blue': ' sad',
 'bold': ' fearless',
 'bored': ' bored',
 'brave': ' fearless',
 'bright': ' happy',
 'calm': ' safe',
 'capable': ' adequate',
 'certain': ' fearless',
 'challenged': ' attracted',
 'cheated': ' cheated',
 'cherished': ' attached',
 'chicken': ' fearful',
 'comfortable': ' happy',
 'committed': ' focused',
 'common': ' average',
 'compassionate': ' attached',
 



---
# **Part 3**
*   Identifying the common words between two documents.
---




In [87]:
# Converting tokanized word list to set

tokanized_words_without_stopwords_trump_set = set(tokanized_words_without_stopwords_trump)
tokanized_words_without_stopwords_joe_set = set(tokanized_words_without_stopwords_joe)

In [93]:
# Findinng common words between two sets

common_words = tokanized_words_without_stopwords_joe_set.intersection(tokanized_words_without_stopwords_trump_set)
print("Total common words:", len(common_words))


Total common words: 3815



{'funeral',
 'indeed',
 'gloves',
 'officially',
 'videos',
 'blown',
 'warren',
 'car',
 'host',
 '10000',
 'putin’s',
 'ideas',
 'communities',
 'bother',
 'divisions',
 'equality',
 'scum',
 'mexico',
 'personally',
 'write',
 'box',
 'red',
 'boat',
 'office',
 'potential',
 'bidding',
 'only',
 'training',
 'chinese',
 'comprehend',
 'accomplishments',
 'ring',
 'restrictions',
 'our',
 'ccp',
 'false',
 'demands',
 'ok',
 'needs',
 'iraq',
 'experience',
 'top',
 'green',
 'nailed',
 'physically',
 'upcoming',
 'society',
 'did',
 'cold',
 '“we',
 '“trump',
 'followers',
 'encouraged',
 'instagram',
 'depression',
 'ps',
 'kong',
 'attack',
 'fort',
 'violence',
 '44th',
 'founded',
 'intelligence',
 'hopefully',
 'love',
 'allowed',
 'too',
 'eat',
 'supremacists',
 'missed',
 'progressives',
 'trouble',
 'duke',
 '68',
 'use',
 'mitch',
 'et',
 'remove',
 'driver',
 'goodyear',
 'problem',
 'jim',
 'msm',
 'florida',
 'fewer',
 'thankyou',
 'pulled',
 'seanhannity',
 'dejoy',
 

In [94]:
print("Common words:")
print()
common_words

Common words:



{'funeral',
 'indeed',
 'gloves',
 'officially',
 'videos',
 'blown',
 'warren',
 'car',
 'host',
 '10000',
 'putin’s',
 'ideas',
 'communities',
 'bother',
 'divisions',
 'equality',
 'scum',
 'mexico',
 'personally',
 'write',
 'box',
 'red',
 'boat',
 'office',
 'potential',
 'bidding',
 'only',
 'training',
 'chinese',
 'comprehend',
 'accomplishments',
 'ring',
 'restrictions',
 'our',
 'ccp',
 'false',
 'demands',
 'ok',
 'needs',
 'iraq',
 'experience',
 'top',
 'green',
 'nailed',
 'physically',
 'upcoming',
 'society',
 'did',
 'cold',
 '“we',
 '“trump',
 'followers',
 'encouraged',
 'instagram',
 'depression',
 'ps',
 'kong',
 'attack',
 'fort',
 'violence',
 '44th',
 'founded',
 'intelligence',
 'hopefully',
 'love',
 'allowed',
 'too',
 'eat',
 'supremacists',
 'missed',
 'progressives',
 'trouble',
 'duke',
 '68',
 'use',
 'mitch',
 'et',
 'remove',
 'driver',
 'goodyear',
 'problem',
 'jim',
 'msm',
 'florida',
 'fewer',
 'thankyou',
 'pulled',
 'seanhannity',
 'dejoy',
 